File: movie_emoji.rb
#!/usr/bin/env ruby
# movie_emoji.rb
# encoding: UTF-8

require 'json'
require 'fileutils'
require 'timeout'

COLORS = {
  reset: "\e[0m",
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  cyan: "\e[96m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

MOVIES = {
  'easy' => [
    { emoji: '🦁👑', title: 'Король Лев', hint: 'Мультфильм Диснея' },
    { emoji: '🚀🌌', title: 'Звёздные войны', hint: 'Космическая сага' },
    { emoji: '🧙‍♂️💍', title: 'Властелин колец', hint: 'Фэнтези-трилогия' },
    { emoji: '🦇🃏', title: 'Тёмный рыцарь', hint: 'Фильм о Бэтмене' },
    { emoji: '🤖🔫', title: 'Терминатор', hint: 'Культовый фантастический боевик' },
    { emoji: '🦈🌊', title: 'Челюсти', hint: 'Фильм ужасов о большой белой акуле' },
    { emoji: '👽👾', title: 'Инопланетянин', hint: 'Стивен Спилберг' },
    { emoji: '🚢💖', title: 'Титаник', hint: 'Романтическая драма' },
    { emoji: '🐉⚔️', title: 'Гарри Поттер', hint: 'Фильм о волшебнике' },
    { emoji: '🧟‍♂️🧠', title: 'Зомбиленд', hint: 'Кинокомедия про зомби' },
  ],
  'medium' => [
    { emoji: '🔫🐟', title: 'Крепкий орешек', hint: 'Боевик с Брюсом Уиллисом' },
    { emoji: '👻🏠', title: 'Оно', hint: 'Фильм ужасов о клоуне' },
    { emoji: '🧜‍♀️🧽', title: 'Губка Боб', hint: 'Мультфильм о морских обитателях' },
    { emoji: '🐺📈', title: 'Волк с Уолл-стрит', hint: 'Драма о финансисте' },
    { emoji: '🌪️🏠', title: 'Унесённые ветром', hint: 'Классическая мелодрама' },
    { emoji: '🐧🎬', title: 'Мадагаскар', hint: 'Мультфильм о животных' },
    { emoji: '🤡🎈', title: 'Оно', hint: 'Фильм ужасов' },
    { emoji: '🧛‍♂️🌙', title: 'Дракула', hint: 'Готический фильм ужасов' },
    { emoji: '🚗💨', title: 'Форсаж', hint: 'Скоростной боевик' },
    { emoji: '🔪😱', title: 'Пила', hint: 'Фильм ужасов' },
  ],
  'hard' => [
    { emoji: '🦑🐟', title: 'В поисках Немо', hint: 'Мультфильм о рыбках' },
    { emoji: '🐭🍝', title: 'Рататуй', hint: 'Мультфильм о крысе-поваре' },
    { emoji: '🎪🎭', title: 'Безумный Макс', hint: 'Постапокалиптический боевик' },
    { emoji: '🚲🎬', title: 'Велосипедист', hint: 'Драма о гонщике' },
    { emoji: '🐺📖', title: 'Американский психопат', hint: 'Психологический триллер' },
    { emoji: '👨‍🦳🧠', title: 'Игры разума', hint: 'Драма о гениальном математике' },
    { emoji: '🕵️‍♂️🔍', title: 'Шерлок Холмс', hint: 'Детективный фильм' },
    { emoji: '🧛‍♂️❤️', title: 'Сумерки', hint: 'Фильм о вампирах' },
    { emoji: '🦍🏢', title: 'Кинг-Конг', hint: 'Приключенческий фильм' },
    { emoji: '👽🌍', title: 'День независимости', hint: 'Фантастический фильм' },
  ]
}

class MovieEmojiGame
  attr_reader :level, :movies, :max_movies, :score, :correct, :skipped,
              :hint_used, :time_limit, :stats, :stats_file

  def initialize(level = 'easy')
    @level = level
    @movies = MOVIES[level]
    @max_movies = 10
    @score = 0
    @correct = 0
    @skipped = 0
    @hint_used = 0
    @stats_file = File.join(Dir.home, '.movie_emoji_stats.json')
    load_stats
    @time_limit = { 'easy' => 60, 'medium' => 45, 'hard' => 30 }[level]
  end

  def load_stats
    if File.exist?(@stats_file)
      @stats = JSON.parse(File.read(@stats_file))
    else
      @stats = { 'games' => 0, 'best_score' => 0, 'total_movies' => 0 }
    end
  end

  def save_stats
    File.write(@stats_file, JSON.pretty_generate(@stats))
  end

  def timer_input(prompt, timeout)
    print colorize(prompt, :bold)
    answer = nil
    begin
      Timeout.timeout(timeout) do
        answer = STDIN.gets.chomp
      end
    rescue Timeout::Error
      answer = nil
    end
    answer
  end

  def show_hint(movie)
    puts colorize("Подсказка: первая буква '#{movie[:title][0]}', всего #{movie[:title].length} букв", :green)
  end

  def play_round(movie)
    puts colorize("\n🎬 Эмодзи: #{movie[:emoji]}", :cyan)
    puts colorize("Подсказка: #{movie[:hint]}", :blue)
    puts 'Введите название фильма, ? для подсказки, pass для пропуска, quit для выхода.'

    start = Time.now
    answer = timer_input('Ваш ответ: ', @time_limit)
    elapsed = Time.now - start

    if answer.nil?
      puts colorize('⏰ Время вышло!', :red)
      @skipped += 1
      puts colorize("Загаданный фильм: #{movie[:title]}", :yellow)
      return
    end
    if answer == 'quit'
      puts 'Выход.'
      save_stats
      exit
    end
    if answer == '?'
      if @hint_used < 1
        show_hint(movie)
        @hint_used += 1
        @score = [0, @score - 5].max
        play_round(movie)
        return
      else
        puts colorize('Подсказка уже использована.', :yellow)
        play_round(movie)
        return
      end
    end
    if answer == 'pass'
      @skipped += 1
      puts colorize("Загаданный фильм: #{movie[:title]}", :yellow)
      return
    end
    if answer.casecmp(movie[:title]).zero?
      @correct += 1
      points = [1, (10 - elapsed / 10).to_i].max
      @score += points
      puts colorize("✅ Верно! +#{points} очков. Время: #{elapsed.round(1)} сек", :green)
    else
      @skipped += 1
      puts colorize("❌ Неверно. Загаданный фильм: #{movie[:title]}", :red)
    end
  end

  def play
    puts colorize('🎬 Добро пожаловать в игру "Угадай фильм по эмодзи"!', :bold)
    puts "Уровень: #{@level}, лимит времени: #{@time_limit} сек."
    puts 'Цель: угадать фильм по набору эмодзи.'
    puts 'Вводите название, используйте ? для подсказки, pass для пропуска, quit для выхода.'

    @movies.shuffle!
    total = [@max_movies, @movies.size].min
    (0...total).each do |i|
      puts colorize("\nФильм #{i+1}/#{total}", :blue)
      play_round(@movies[i])
    end
    puts colorize("\n🏁 Игра завершена!", :bold)
    puts "  Угадано фильмов: #{@correct}"
    puts "  Пропущено: #{@skipped}"
    puts "  Использовано подсказок: #{@hint_used}"
    puts "  Счёт: #{@score}"
    @stats['games'] += 1
    if @score > @stats['best_score']
      @stats['best_score'] = @score
      puts colorize('🏆 Новый рекорд!', :green)
    end
    @stats['total_movies'] += @correct
    save_stats
    puts colorize("Лучший результат: #{@stats['best_score']}", :yellow)
  end
end

def main
  level = 'easy'
  show_stats = false
  reset_stats = false
  ARGV.each do |arg|
    case arg
    when 'easy', 'medium', 'hard'
      level = arg
    when '-s', '--stats'
      show_stats = true
    when '-r', '--reset'
      reset_stats = true
    when '-h', '--help'
      puts 'Usage: ruby movie_emoji.rb [easy|medium|hard] [-s] [-r]'
      return
    end
  end
  if reset_stats
    f = File.join(Dir.home, '.movie_emoji_stats.json')
    File.delete(f) if File.exist?(f)
    puts 'Статистика сброшена.'
    return
  end
  if show_stats
    f = File.join(Dir.home, '.movie_emoji_stats.json')
    if File.exist?(f)
      stats = JSON.parse(File.read(f))
      puts colorize('📊 Статистика:', :bold)
      puts "  Сыграно игр: #{stats['games']}"
      puts "  Лучший счёт: #{stats['best_score']}"
      puts "  Всего угадано фильмов: #{stats['total_movies']}"
    else
      puts 'Статистика пуста.'
    end
    return
  end
  game = MovieEmojiGame.new(level)
  game.play
end

main if __FILE__ == $0

File: movie_emoji.py
# movie_emoji.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import os
import random
import json
import time
import threading
from pathlib import Path

# ANSI-цвета
COLORS = {
    'reset': '\033[0m',
    'green': '\033[92m',
    'red': '\033[91m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'cyan': '\033[96m',
    'bold': '\033[1m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

# Словарь фильмов: (эмодзи, название, подсказка)
MOVIES = {
    'easy': [
        ('🦁👑', 'Король Лев', 'Мультфильм Диснея'),
        ('🚀🌌', 'Звёздные войны', 'Космическая сага'),
        ('🧙‍♂️💍', 'Властелин колец', 'Фэнтези-трилогия'),
        ('🦇🃏', 'Тёмный рыцарь', 'Фильм о Бэтмене'),
        ('🤖🔫', 'Терминатор', 'Культовый фантастический боевик'),
        ('🦈🌊', 'Челюсти', 'Фильм ужасов о большой белой акуле'),
        ('👽👾', 'Инопланетянин', 'Стивен Спилберг'),
        ('🚢💖', 'Титаник', 'Романтическая драма'),
        ('🐉⚔️', 'Гарри Поттер', 'Фильм о волшебнике'),
        ('🧟‍♂️🧠', 'Зомбиленд', 'Кинокомедия про зомби'),
    ],
    'medium': [
        ('🔫🐟', 'Крепкий орешек', 'Боевик с Брюсом Уиллисом'),
        ('👻🏠', 'Оно', 'Фильм ужасов о клоуне'),
        ('🧜‍♀️🧽', 'Губка Боб', 'Мультфильм о морских обитателях'),
        ('🐺📈', 'Волк с Уолл-стрит', 'Драма о финансисте'),
        ('🌪️🏠', 'Унесённые ветром', 'Классическая мелодрама'),
        ('🐧🎬', 'Мадагаскар', 'Мультфильм о животных'),
        ('🤡🎈', 'Оно', 'Фильм ужасов'),
        ('🧛‍♂️🌙', 'Дракула', 'Готический фильм ужасов'),
        ('🚗💨', 'Форсаж', 'Скоростной боевик'),
        ('🔪😱', 'Пила', 'Фильм ужасов'),
    ],
    'hard': [
        ('🦑🐟', 'В поисках Немо', 'Мультфильм о рыбках'),
        ('🐭🍝', 'Рататуй', 'Мультфильм о крысе-поваре'),
        ('🎪🎭', 'Безумный Макс', 'Постапокалиптический боевик'),
        ('🚲🎬', 'Велосипедист', 'Драма о гонщике'),
        ('🐺📖', 'Американский психопат', 'Психологический триллер'),
        ('👨‍🦳🧠', 'Игры разума', 'Драма о гениальном математике'),
        ('🕵️‍♂️🔍', 'Шерлок Холмс', 'Детективный фильм'),
        ('🧛‍♂️❤️', 'Сумерки', 'Фильм о вампирах'),
        ('🦍🏢', 'Кинг-Конг', 'Приключенческий фильм'),
        ('👽🌍', 'День независимости', 'Фантастический фильм'),
    ]
}

class MovieEmojiGame:
    def __init__(self, level='easy'):
        self.level = level
        self.movies = MOVIES[level]
        self.max_movies = 10
        self.score = 0
        self.correct = 0
        self.skipped = 0
        self.hint_used = 0
        self.stats_file = Path.home() / '.movie_emoji_stats.json'
        self.load_stats()
        self.time_limit = {'easy': 60, 'medium': 45, 'hard': 30}[level]

    def load_stats(self):
        if self.stats_file.exists():
            with open(self.stats_file, 'r') as f:
                self.stats = json.load(f)
        else:
            self.stats = {'games': 0, 'best_score': 0, 'total_movies': 0}

    def save_stats(self):
        with open(self.stats_file, 'w') as f:
            json.dump(self.stats, f, indent=2)

    def timer_input(self, prompt, timeout):
        print(colorize(prompt, 'bold'), end='', flush=True)
        user_input = ['']
        def get_input():
            try:
                user_input[0] = sys.stdin.readline().strip()
            except:
                pass
        thread = threading.Thread(target=get_input)
        thread.daemon = True
        thread.start()
        start = time.time()
        while time.time() - start < timeout:
            if user_input[0] != '':
                return user_input[0]
            remaining = int(timeout - (time.time() - start))
            sys.stdout.write(f"\r{colorize(f'Осталось времени: {remaining} сек', 'yellow')}")
            sys.stdout.flush()
            time.sleep(1)
        return None

    def show_hint(self, movie):
        # Показываем первую букву или количество букв
        title = movie[1]
        print(colorize(f"Подсказка: первая буква '{title[0]}', всего {len(title)} букв", 'green'))

    def play_round(self, movie):
        emoji, title, hint = movie
        print(colorize(f"\n🎬 Эмодзи: {emoji}", 'cyan'))
        print(colorize(f"Подсказка: {hint}", 'blue'))
        print("Введите название фильма, ? для подсказки, pass для пропуска, quit для выхода.")

        start_time = time.time()
        answer = self.timer_input("Ваш ответ: ", self.time_limit)
        elapsed = time.time() - start_time

        if answer is None:
            print(colorize("⏰ Время вышло!", 'red'))
            self.skipped += 1
            print(colorize(f"Загаданный фильм: {title}", 'yellow'))
            return

        if answer == 'quit':
            print("Выход.")
            self.save_stats()
            sys.exit(0)
        if answer == '?':
            if self.hint_used < 1:
                self.show_hint(movie)
                self.hint_used += 1
                self.score = max(0, self.score - 5)
                self.play_round(movie)
                return
            else:
                print(colorize("Подсказка уже использована.", 'yellow'))
                self.play_round(movie)
                return
        if answer == 'pass':
            self.skipped += 1
            print(colorize(f"Загаданный фильм: {title}", 'yellow'))
            return

        if answer.lower() == title.lower():
            self.correct += 1
            points = max(1, int(10 - elapsed / 10))
            self.score += points
            print(colorize(f"✅ Верно! +{points} очков. Время: {elapsed:.1f} сек", 'green'))
        else:
            self.skipped += 1
            print(colorize(f"❌ Неверно. Загаданный фильм: {title}", 'red'))

    def play(self):
        print(colorize("🎬 Добро пожаловать в игру 'Угадай фильм по эмодзи'!", 'bold'))
        print(f"Уровень: {self.level}, лимит времени: {self.time_limit} сек.")
        print("Цель: угадать фильм по набору эмодзи.")
        print("Вводите название, используйте ? для подсказки, pass для пропуска, quit для выхода.\n")

        random.shuffle(self.movies)
        total = min(self.max_movies, len(self.movies))
        for i in range(total):
            print(colorize(f"\nФильм {i+1}/{total}", 'blue'))
            self.play_round(self.movies[i])

        print(colorize("\n🏁 Игра завершена!", 'bold'))
        print(f"  Угадано фильмов: {self.correct}")
        print(f"  Пропущено: {self.skipped}")
        print(f"  Использовано подсказок: {self.hint_used}")
        print(f"  Счёт: {self.score}")
        self.stats['games'] += 1
        if self.score > self.stats['best_score']:
            self.stats['best_score'] = self.score
            print(colorize("🏆 Новый рекорд!", 'green'))
        self.stats['total_movies'] += self.correct
        self.save_stats()
        print(colorize(f"Лучший результат: {self.stats['best_score']}", 'yellow'))

def main():
    level = 'easy'
    show_stats = False
    reset_stats = False
    args = sys.argv[1:]
    for arg in args:
        if arg in ['easy', 'medium', 'hard']:
            level = arg
        elif arg == '-s' or arg == '--stats':
            show_stats = True
        elif arg == '-r' or arg == '--reset':
            reset_stats = True
        elif arg == '-h' or arg == '--help':
            print("Usage: movie_emoji.py [easy|medium|hard] [-s] [-r]")
            return
    if reset_stats:
        stats_file = Path.home() / '.movie_emoji_stats.json'
        if stats_file.exists():
            stats_file.unlink()
        print("Статистика сброшена.")
        return
    if show_stats:
        stats_file = Path.home() / '.movie_emoji_stats.json'
        if stats_file.exists():
            with open(stats_file, 'r') as f:
                stats = json.load(f)
                print(colorize("📊 Статистика:", 'bold'))
                print(f"  Сыграно игр: {stats['games']}")
                print(f"  Лучший счёт: {stats['best_score']}")
                print(f"  Всего угадано фильмов: {stats['total_movies']}")
        else:
            print("Статистика пуста.")
        return
    game = MovieEmojiGame(level)
    game.play()

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print(colorize("\nИгра прервана.", 'yellow'))
        sys.exit(0)
